pkg/daemon/command/command.go

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2022. Nydus Developers. All rights reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package command
+
+import (
+	"fmt"
+	"reflect"
+	"strconv"
+)
+
+type Opt = func(cmd *DaemonCommand)
+
+// Define how to build a command line to start a nydusd daemon
+type DaemonCommand struct {
+	// "singleton" "fuse"
+	Mode string `type:"subcommand"`
+	// "blobcache" "fscache" "virtiofs"
+	FscacheDriver  string `type:"param" name:"fscache"`
+	FscacheThreads string `type:"param" name:"fscache-threads"`
+	Upgrade        bool   `type:"flag" name:"upgrade" default:""`
+	ThreadNum      string `type:"param" name:"thread-num"`
+	Config         string `type:"param" name:"config"`
+	Bootstrap      string `type:"param" name:"bootstrap"`
+	Mountpoint     string `type:"param" name:"mountpoint"`
+	APISock        string `type:"param" name:"apisock"`
+	LogLevel       string `type:"param" name:"log-level"`
+	Supervisor     string `type:"param" name:"supervisor"`
+	LogFile        string `type:"param" name:"log-file"`
+}
+
+// Build exec style command line
+func BuildCommand(opts []Opt) []string {
+	var cmd DaemonCommand
+	var subcommand string
+
+	for _, o := range opts {
+		o(&cmd)
+	}
+
+	args := make([]string, 0, 32)
+	t := reflect.TypeOf(cmd)
+	v := reflect.ValueOf(cmd)
+
+	for i := 0; i < t.NumField(); i++ {
+		tag := t.Field(i).Tag
+		argType := tag.Get("type")
+
+		switch argType {
+		case "param":
+			// Zero value will be skipped appending to command line
+			if v.Field(i).IsZero() {
+				continue
+			}
+
+			value := v.Field(i).Interface()
+
+			pair := []string{fmt.Sprintf("--%s", tag.Get("name")), fmt.Sprintf("%s", value)}
+			args = append(args, pair...)
+		case "subcommand":
+			// Zero value will be skipped appending to command line
+			if v.Field(i).IsZero() {
+				continue
+			}
+			subcommand = v.Field(i).String()
+		case "flag":
+			kind := v.Field(i).Kind()
+
+			if kind != reflect.Bool {
+				panic("flag must be boolean")
+			}
+
+			v := v.Field(i).Bool()
+
+			if v {
+				flag := fmt.Sprintf("--%s", tag.Get("name"))
+				args = append(args, flag)
+			} else {
+				continue
+			}
+		}
```

Comment:
Better to handle `default` and give an error message, in case of an invalid option field is added.

Reply:
good point

Reply:
fixed